File: LWHeroOC/Classes/HeroTargetState.m
//
//  HeroTargetState.m
//  Hero-ObjectiveC
//
//  Created by luca.li on 2017/1/22.
//  Copyright © 2017年 luca. All rights reserved.
//

#import "HeroTargetState.h"
#import "HeroModifier.h"

@interface HeroTargetState ()

// @[   @{@"modifier1" : modifier1},
//      @{@"modifier2" : modifier2} ]
@property (nonatomic, copy) NSMutableArray <NSMutableDictionary*> *custom;

@end

@implementation HeroTargetState

- (instancetype)initWithModifiers:(NSArray *)modifiers {
    if (self = [super init]) {
        [self appendContentsOfModifiers:modifiers];
    }
    return self;
}

- (void)appendContentsOfModifiers:(NSArray <HeroModifier *> *)modifiers {
    for (HeroModifier *modifier in modifiers) {
        modifier.apply(self);
    }
}

- (id)customItemOfKey:(NSString *)key {
    __block id modifier = nil;
    [self.custom enumerateObjectsUsingBlock:^(NSMutableDictionary * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj objectForKey:key]) {
            modifier = [obj objectForKey:key];
        }
    }];
    
    return modifier;
}

- (void)setCustomItemOfKey:(NSString *)key value:(id)value {
    if (!self.custom) {
        self.custom = [NSMutableArray array];
    }
    
    __block BOOL contain = NO;
    __block NSInteger index = 0;
    [self.custom enumerateObjectsUsingBlock:^(NSMutableDictionary * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj objectForKey:key]) {
            contain = YES;
            index = idx;
            *stop = YES;
        }
    }];
    
    if (contain) {
        [self.custom removeObjectAtIndex:index];
    }
    
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setValue:value forKey:key];
    [self.custom addObject:dict];
}

#pragma mark - Getter

- (NSNumber *)delay {
    //Default 0
    if (!_delay) {
        return [NSNumber numberWithInteger:0];
    }
    return _delay;
}

#pragma mark - NSCopying

- (id)copyWithZone:(NSZone *)zone {
    HeroTargetState *copyState = [[self class] allocWithZone:zone];
    copyState->_opacity = _opacity;
    copyState->_cornerRadius = _cornerRadius;
    copyState->_position = _position;
    copyState->_size = _size;
    copyState->_transform = _transform;
    copyState->_spring = _spring;
    copyState->_delay = _delay;
    copyState->_duration = _duration;
    copyState->_timingFunction = _timingFunction;
    copyState->_arc =_arc;
    copyState->_zPosition = _zPosition;
    copyState->_zPositionIfMatched = _zPositionIfMatched;
    copyState->_source = _source;
    copyState->_cascade = _cascade;
    
    copyState->_ignoreSubviewModifiers = _ignoreSubviewModifiers;
    
    copyState->_custom = [NSMutableArray array];
    copyState->_custom = _custom;
    
    return copyState;
}

@end


@implementation HeroTargetState (ExpressibleByArrayLiteral)

- (instancetype)initWithArrayLiteralElements:(NSArray <HeroModifier *> *)elements {
    if (self = [super init]) {
        [self appendContentsOfModifiers:elements];
    }
    return self;
}

@end

File: LWHero_swift/LWHero/Classes/HeroTargetState.swift
//
//  HeroTargetState.swift
//  LWHero
//
//  Created by Swift Port
//  Copyright © 2025. All rights reserved.
//

import Foundation
import UIKit
import QuartzCore

public class HeroTargetState: NSObject, NSCopying {
    // MARK: - Properties

    public var opacity: CGFloat?
    public var cornerRadius: CGFloat?
    public var position: CGPoint?
    public var size: CGSize?
    public var transform: CATransform3D?
    public var spring: (stiffness: CGFloat, damping: CGFloat)?
    public var delay: TimeInterval?
    public var duration: TimeInterval?
    public var timingFunction: CAMediaTimingFunction?
    public var arc: CGFloat?
    public var zPosition: CGFloat?
    public var zPositionIfMatched: CGFloat?
    public var source: String?
    public var cascade: (delta: TimeInterval, direction: CascadeDirection, delayMatchedViews: Bool)?
    public var useGlobalCoordinateSpace: Bool = false
    public var ignoreSubviewModifiers: Bool?

    public private(set) var custom: [[String: Any]] = []

    // MARK: - Initialization

    public override init() {
        super.init()
    }

    public init(modifiers: [HeroModifier]) {
        super.init()
        append(modifiers: modifiers)
    }

    // MARK: - Modifier Management

    public func append(modifiers: [HeroModifier]) {
        for modifier in modifiers {
            modifier.apply(self)
        }
    }

    // MARK: - Custom Items

    public func customItem(forKey key: String) -> Any? {
        for dict in custom {
            if let value = dict[key] {
                return value
            }
        }
        return nil
    }

    public func setCustomItem(_ value: Any?, forKey key: String) {
        if let value = value {
            var found = false
            for (index, dict) in custom.enumerated() {
                if dict[key] != nil {
                    custom[index][key] = value
                    found = true
                    break
                }
            }
            if !found {
                custom.append([key: value])
            }
        } else {
            custom = custom.map { dict in
                var mutableDict = dict
                mutableDict.removeValue(forKey: key)
                return mutableDict
            }.filter { !$0.isEmpty }
        }
    }

    // MARK: - NSCopying

    public func copy(with zone: NSZone? = nil) -> Any {
        let copy = HeroTargetState()
        copy.opacity = opacity
        copy.cornerRadius = cornerRadius
        copy.position = position
        copy.size = size
        copy.transform = transform
        copy.spring = spring
        copy.delay = delay
        copy.duration = duration
        copy.timingFunction = timingFunction
        copy.arc = arc
        copy.zPosition = zPosition
        copy.zPositionIfMatched = zPositionIfMatched
        copy.source = source
        copy.cascade = cascade
        copy.useGlobalCoordinateSpace = useGlobalCoordinateSpace
        copy.ignoreSubviewModifiers = ignoreSubviewModifiers
        copy.custom = custom
        return copy
    }
}

// MARK: - Array Literal

extension HeroTargetState: ExpressibleByArrayLiteral {
    public convenience init(arrayLiteral elements: HeroModifier...) {
        self.init(modifiers: elements)
    }
}
